atom/browser/api/atom_api_web_contents.cc

```diff
@@ -411,14 +411,18 @@ WebContents::~WebContents() {
     if (type_ == WEB_VIEW)
       guest_delegate_->Destroy();
 
-    // The WebContentsDestroyed will not be called automatically because we
-    // unsubscribe from webContents before destroying it. So we have to manually
-    // call it here to make sure "destroyed" event is emitted.
     RenderViewDeleted(web_contents()->GetRenderViewHost());
-    WebContentsDestroyed();
+    DestroyWebContents();
   }
 }
 
+void WebContents::DestroyWebContents() {
+  // This event is only for internal use, which is emitted when WebContents is
+  // being destroyed.
+  Emit("will-destroy");
+  CommonWebContentsDelegate::DestroyWebContents();
```

Comment:
`CommonWebContentsDelegate::DestroyWebContents` should be made a `protected` method and renamed to something else like `ResetWebContents`, having a non-virtual method with the same name is not good practice.